Remora.Commands/Signatures/IParameterShape.cs

```diff
@@ -56,6 +53,31 @@ public interface IParameterShape
     /// </summary>
     string Description { get; }
 
+    /// <summary>
+    /// Gets a list of attributes that apply to this parameter.
+    /// </summary>
+    IReadOnlyList<Attribute> Attributes { get; }
+
+    /// <summary>
+    /// Gets a list of conditions that apply to this parameter.
+    /// </summary>
+    IReadOnlyList<ConditionAttribute> Conditions { get; }
+
+    /// <summary>
+    /// Gets the type the parameter represents, or otherwise should be parsed as.
+    /// </summary>
+    Type ParameterType { get; }
+
+    /// <summary>
+    /// Gets a value indicating whether the parameter's type allows for null values.
+    /// </summary>
+    bool IsNullable { get; }
+
+    /// <summary>
+    /// Gets the index of the parameter in the command's signature.
+    /// </summary>
+    int ParameterIndex { get; }
```

Comment:
Does the index of the parameter need to be part of its shape? Typically, this information would be available to the user when dealing with shapes and is generally not useful for matching input to the shape of the parameter itself.